### DISCUSSION TITLE
Automatic verification and updates of Quarkus

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#8985] [Ideas]
We have consistently had challenges with upgrading to new WildFly releases, which results in a large gap between when WildFly is released, and a Keycloak release with the upgrade to the new WildFly. To illustrate this WildFly 24 was released on June 17th, WildFly 25 was released on October 5th, and we still don't have a Keycloak release available based on either.

Purely from a perspective of CVEs in dependencies inherited from WildFly this is not acceptable. In addition it is taking a lot of resources to manage these upgrades.

With Keycloak.X and Quarkus the aim is to continuously upgrade, with Keycloak releases being done as close as possible to a Quarkus release as possible. This means we need to verify Quarkus continuously has development happens, rather than at the time of a release.

One issue with that is we want our main branch to be updated only when an actual Quarkus release has happened, and there can be incompatibilities between the latest Quarkus release and Quarkus main branch.

The solution I'd like to propose for this is creating a GitHub workflow that can automate the following:

* Maintain a "quarkus-update" branch that is automatically rebased with the main branch as PRs are merged
* Update Quarkus in the "quarkus-update" branch with the latest snapshot release of Quarkus
* Detect a new Final release of Quarkus and open a PR against main branch

The workflow should at least run nightly, but could run more frequently than that. The reason behind the "quarkus-update" branch is that if there are changes required we can detect this and send a PR to resolve these issues to the "quarkus-update" branch without breaking the main branch which needs to remain compatible with the latest Quarkus release.

One issue that we can't solve right now, is that we can't inherit dependency versions from the Quarkus BOM, which requires manually updating dependency versions in https://github.com/keycloak/keycloak/blob/main/quarkus/pom.xml. This is something we need to live with for now until we can remove support for WildFly, and potentially also move adapters into a separate repository. In the long run though versions of dependencies inherited from Quarkus should set by adding the Quarkus BOM to the Keycloak parent.

**Comment by stianst** [comment]
@andreaTP @DGuhr @pedroigor FIY

**Reply by andreaTP** [reply]
I think that most of those requirements are met by the https://github.com/quarkusio/quarkus-ecosystem-ci project directly targeting `main` in the Keycloak repository:

 - Test against Quarkus Nightlies
 - Keep testing against a Quarkus LTS (can be achieved using the [`alternatives`](https://github.com/quarkusio/quarkus-ecosystem-ci/blob/2d9ec9222999a88716c98ff09d9a41a861793793/kogito-quarkus/context.yaml#L8) )
 
 Not covered is the automation to detect a new Final release and open the relevant PR (should we consider using something like dependabot for this use-case?).
 
 > we can't inherit dependency versions from the Quarkus BOM
 
This is correct and is a major risk factor.

**Reply by stianst** [reply]
quarkus-ecosystem-ci assumes it is an extension registered in the Quarkus ecosystem, which we are not, and will not be. I'm unsure what that means though, but scanning through the quarkus-ecosystem-ci readme there's quite a bit of things that makes that assumption.

**Reply by stianst** [reply]
If you target the main repository directly there is no opportunity to make changes needed due to expected breaking changes in Quarkus, and also to upgrade dependencies (until we can inherit directly from Quarkus BOM).

**Reply by andreaTP** [reply]
regarding `quarkus-ecosystem-ci` I hope someone like @gsmet or @maxandersen can weight into this conversation.

**Reply by andreaTP** [reply]
not targeting the `main` branch might require some automation or process to handle it ... but this is going to be unavoidable unless we reach full compatibility ...

**Reply by gsmet** [reply]
Hey there.

The Quarkus ecosystem CI is very customizable so I think we could make it work for you and it definitely has a lot of value.

As for tracking our versions, keep in mind that you can restrict the scope of Dependabot so, typically, you could enable it only for Quarkus if you don't want to enable it for the rest of your dependencies.

**Reply by maxandersen** [reply]
Ecosystem is absolutely not limited to quarkiverse. The default scripts assume it's an extension but you can customize what keycloak side does. 

For core bom syncing I'm wondering if we can explore other ways. Is there any other kind of artifact you could depend on we could generate ?

**Reply by stianst** [reply]
Thanks, that's good to know that we can rely on the Quarkus ecosystem CI system.

**Reply by stianst** [reply]
> For core bom syncing I'm wondering if we can explore other ways. Is there any other kind of artifact you could depend on we could generate ?

I don't think so. The problem we have right now is that we define versions of some dependencies in the top-level parent pom, which sets versions used by WF distribution, adapters, testsuite, etc.. Right now we can't use the versions defined in the Quarkus BOM everywhere, and nor can we override the versions from the BOM in the Quarkus modules I think. As we get rid of WildFly distribution, and also refactor a bit around adapters we should be able to just import the Quarkus BOM at the top-level.

**Reply by andreaTP** [reply]
I prepared a stub of the technical integration with `quarkus-ecosystem-ci` here: https://github.com/keycloak/keycloak/pull/8971
That PR can be used as a starting point to cover the additional requirements @stianst if you agree on moving forward in this direction.

@gsmet and @maxandersen thanks a lot for your input!

**Reply by andreaTP** [reply]
@stianst I was considering the requirement of having a separate `quarkus-update` branch and I would like to know if we can start small and eventually implement it when we feel the need, here you have the main motivations:

 - All the merge actions are requiring quite high privileges and they are all "third-party"
 - Quarkus, so far, demonstrated to be pretty much stable or, alternatively, really quick in actioning any issue found
 - There is no "immediate need" for heavy automation since the process is not really automated yet
 
Regarding the update of the dependencies, I do believe that we can simply exclude it at first with a tuned selective update (like the one implemented [here](https://github.com/keycloak/keycloak/pull/8971/files#diff-1afa90ff1b1076491994d11cc19d79dd2e0f58aa0db6867b0c6280f3442ed03bR9)) and improve on it over time.

All in all I feel that we can keep your initial indication as a final goal and progress toward it with smaller steps.

**Reply by stianst** [reply]
From looking at upgrades of Quarkus in the past, most upgrades have included some code changes on our end. The reasoning for a separate branch is that I want to have the upgrade ready and prepared in advance of the Quarkus final release, as we want to release as close to the Quarkus release as we can.

**Reply by andreaTP** [reply]
@stianst here you have a prototype for a possible solution to the separate branch:

https://github.com/andreaTP/keycloak/blob/test-quarkus-branch/.github/workflows/quarkus-branch.yaml

Since those actions are third-party I would suggest a code review(I did it and found the code legit but better to have another couple of eyes) + forking under the keycloak organization.
